File: internal/limiter/distributed/slidingwindowcounter.go
package distributed

import (
	"context"
	"fmt"
	"time"

	"github.com/redis/go-redis/v9"
	"github.com/streamliner/rate-limiter/internal/limiter"
)

// SlidingWindowCounter is a Redis-backed approximate sliding window using two STRING
// counters (current and previous window). O(1) memory; bounded over-admission error.
type SlidingWindowCounter struct {
	spec  limiter.Spec
	rdb   *redis.Client
	clock limiter.Clock
}

func NewSlidingWindowCounter(spec limiter.Spec, rdb *redis.Client, clock limiter.Clock) *SlidingWindowCounter {
	return &SlidingWindowCounter{spec: spec, rdb: rdb, clock: clock}
}

func (swc *SlidingWindowCounter) Allow(ctx context.Context, key string, cost int64) (limiter.Decision, error) {
	if cost <= 0 {
		cost = 1
	}

	now := swc.clock.Now()
	nowMs := now.UnixMilli()
	windowMs := swc.spec.Window.Milliseconds()
	// window_start_ms: floor to the nearest window boundary
	windowStartMs := (nowMs / windowMs) * windowMs
	// TTL is 2× window so the previous-window key stays alive while it is still needed.
	ttlMs := windowMs * 2

	// Both keys must share the same hash tag {key} so they land in one Redis Cluster slot.
	currKey := fmt.Sprintf("ratelimit:{%s}:swc:%d", key, windowStartMs)
	prevKey := fmt.Sprintf("ratelimit:{%s}:swc:%d", key, windowStartMs-windowMs)

	res, err := runScript(ctx, swc.rdb, slidingWindowCounterScript,
		[]string{currKey, prevKey},
		swc.spec.Limit, windowMs, nowMs, cost, windowStartMs, ttlMs,
	)
	if err != nil {
		return limiter.Decision{}, err
	}

	allowed := toInt64(res[0])
	remaining := toInt64(res[1])

	resetAfter := time.Duration(windowMs-(nowMs-windowStartMs)) * time.Millisecond

	if allowed == 1 {
		return limiter.Allowed(swc.spec.Limit, remaining, resetAfter), nil
	}
	return limiter.Denied(swc.spec.Limit, resetAfter, resetAfter), nil
}

File: internal/limiter/distributed/lua/sliding_window_counter.lua
-- sliding_window_counter.lua
-- Approximate sliding window using two fixed-slot counters with weighted blending.
-- O(1) memory per key; bounded error of at most one window's worth of requests.
--
-- KEYS[1]  = current window counter key  (STRING, hash-tagged)
-- KEYS[2]  = previous window counter key (STRING, same hash tag — same Cluster slot)
-- ARGV[1]  = limit             (int)
-- ARGV[2]  = window_ms         (int)
-- ARGV[3]  = now_ms            (int, from Go)
-- ARGV[4]  = cost              (int)
-- ARGV[5]  = window_start_ms   (int, floor(now_ms / window_ms) * window_ms)
-- ARGV[6]  = ttl_ms            (int)
--
-- Returns: {allowed (0|1), remaining (int)}
--
-- The approximation: weight the previous window by the fraction of it still
-- overlapping the current rolling window. At 30% through the current window,
-- 70% of the previous window's requests are still "in range".

local limit            = tonumber(ARGV[1])
local window_ms        = tonumber(ARGV[2])
local now              = tonumber(ARGV[3])
local cost             = tonumber(ARGV[4])
local window_start_ms  = tonumber(ARGV[5])
local ttl              = tonumber(ARGV[6])

local prev = tonumber(redis.call('GET', KEYS[2])) or 0
local curr = tonumber(redis.call('GET', KEYS[1])) or 0

local elapsed  = now - window_start_ms
local overlap  = (window_ms - elapsed) / window_ms
local estimate = prev * overlap + curr

if estimate + cost <= limit then
    redis.call('INCRBY', KEYS[1], cost)
    -- TTL is 2× window so the previous-window key stays alive while it is still needed.
    redis.call('PEXPIRE', KEYS[1], ttl)
    redis.call('PEXPIRE', KEYS[2], ttl)

    local remaining = math.floor(limit - estimate - cost)
    return {1, remaining}
end

return {0, 0}

